platform/sensor-stream-manager/sensor-stream-manager.js

```diff
@@ -1,10 +1,111 @@
+// }
+const sensorStream = {};
+
+// TODO: policy
+const policy = {};
+
+// connect to localhost mqtt broker
+const mqttClient = mqtt.connect("mqtt://localhost");
+mqttClient.on('connect', () => {
+    // subscribe to localhost "gateway-data" topic
+    mqttClient.subscribe("gateway-data", (err) => {
+        if (err) {
+            console.error(`mqtt client subscribe ${mqttTopic} failed.`);
+            console.error(err);
+        } else {
+            console.log(`subscribe to "gateway-data" topic successfully!`);
+        }
+    });
+});
+
+// when sensor stream data is published to "gateway-data" topic
+// SSM will check policy and publish the data to application's topic
+mqttClient.on('message', (topic, message) => {
+    const payload = JSON.parse(message.toString());
+    const sensorId = payload["_meta"]["device_id"];
+    if(sensorId in sensorStream) {
+        for(let topic of sensorStream[sensorId]) {
+            // TODO: check policy
+
+            // publish to application's topic
+            mqttClient.publish(topic, JSON.stringify(payload));
+        }
+    }
+});
+
 messagingService.listenForEvent('connect-to-socket', (message) => {
     const payload = message.data;
     const wsAddress = payload["ws-address"];
+});
+
+// SSM receives application's process instance and metadata from app-manager
+// this listener will store the application's topic and sensor stream requirement
+messagingService.listenForEvent('app-deployment', message => {
+    // appData = {
+    //     "app": {
+    //         "app": process-instance,
+    //         "topic": "app-topic",
+    //         "path": "app-path"
+    //     },
+    //     "metadataPath": "metadata-path"
+    // };
+    let appData = message.data;
+    if(appData["app"] && appData["metadataPath"]) {
+        // load application's metadata
+        let metadata = fs.readJsonSync(appData["metadataPath"]);
+        metadata = metadata["sensorMapping"];
+        let topic = appData["app"]["topic"];
+        // store application's sensor stream requirement in sensorStream
+        for(let ip in metadata) {
+            // store the sensor connected to local gateway
+            if(ip === gatewayIp) {
+                // check if metadata[ip] is an array
+                if(Array.isArray(metadata[ip])) {
+                    for(let id of metadata[ip]) {
+                        if(id in sensorStream) {
+                            sensorStream[id].push(topic);
+                        } else {
+                            sensorStream[id] = [topic];
+                        }
+                    }
+                } else {
+                    // metadata[ip] is a string if it contains only 1 sensor
+                    if(metadata[ip] in sensorStream) {
+                        sensorStream[metadata[ip]].push(topic);
+                    } else {
+                        sensorStream[metadata[ip]] = [topic];
+                    }
+                }
```

Comment:
Why would metadata[ip] be a string if there is only 1 sensor?

Reply:
I thought there is the same [problem](https://github.com/nabeeln7/hoos-nearby/commit/154565faf8f153699fe14d0eb2eb18e25fed932c) in metadata, but after I check the [code](https://github.com/nabeeln7/hoos-nearby/blob/fbc7af1e10d5ea41a287c3dbe569a9e5a9dcf198/auxiliary-device-tools/app-deployer/app-deployer.js#L26), no string problem here. I will fix that.